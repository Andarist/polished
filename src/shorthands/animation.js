// @flow
import polish from '../validation/polish'

/** */
type AnimationProperty = string | number

/**
 * Shorthand for easily setting the animation property. Allows either multiple arrays with animations
 * or a single animation spread over the arguments.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...animation(['rotate', '1s', 'ease-in-out'], ['colorchange', '2s'])
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${animation(['rotate', '1s', 'ease-in-out'], ['colorchange', '2s'])}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'animation': 'rotate 1s ease-in-out, colorchange 2s'
 * }
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...animation('rotate', '1s', 'ease-in-out')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${animation('rotate', '1s', 'ease-in-out')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'animation': 'rotate 1s ease-in-out'
 * }
 */
function animation(
  ...args: Array<Array<AnimationProperty> | AnimationProperty>
) {
  const code = args
    .map(arg => (Array.isArray(arg) ? arg.join(' ') : arg))
    .join(', ')

  return {
    animation: code,
  }
}

export default polish({
  modulePath: 'shorthands/animation',
  types: {
    key: 'animationProperty',
    type: ['array', 'number', 'string'],
    maxLength: 8,
    required: true,
    matchAll: true,
  },
})(animation)
