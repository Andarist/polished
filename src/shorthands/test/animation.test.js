// @flow
import animation from '../animation'

describe('animation', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })
  describe('single mode', () => {
    it('should pass first eight arguments to the CSS', () => {
      expect({
        ...animation(
          'rotate',
          '1s',
          'ease-in-out',
          '0.5s',
          5,
          'reverse',
          'forwards',
          'paused',
        ),
      }).toMatchSnapshot()
    })

    it('should be fine with less than eight arguments', () => {
      expect({ ...animation('rotate', '1s', 'ease-in-out') }).toMatchSnapshot()
    })

    it('should throw a warning if more than eight elements are supplied', () => {
      animation(
        'one',
        'two',
        'three',
        'four',
        'five',
        'six',
        'seven',
        'eight',
        'oops',
      )
      // eslint-disable-next-line no-console
      expect(console.warn.mock.calls).toMatchSnapshot()
    })
  })

  describe('multi mode', () => {
    it('should pass first eight arguments to the CSS in multi mode', () => {
      expect({
        ...animation([
          'rotate',
          '1s',
          'ease-in-out',
          '0.5s',
          5,
          'reverse',
          'forwards',
          'paused',
        ]),
      }).toMatchSnapshot()
    })

    it('should be fine with less than eight arguments', () => {
      expect({
        ...animation(['rotate', '1s', 'ease-in-out']),
      }).toMatchSnapshot()
    })

    it('should be fine with multiple animations', () => {
      expect({
        ...animation(['rotate', '1s', 'ease-in-out'], ['colorchange', '2s']),
      }).toMatchSnapshot()
    })

    it('should throw an error if more than eight elements are supplied in an array', () => {
      expect(() => {
        animation([
          'one',
          'two',
          'three',
          'four',
          'five',
          'six',
          'seven',
          'eight',
          'oops',
        ])
      }).toThrow()
    })

    it('should throw an error if more than eight elements are supplied in a second array', () => {
      expect(() => {
        animation(
          ['rotate'],
          [
            'one',
            'two',
            'three',
            'four',
            'five',
            'six',
            'seven',
            'eight',
            'oops',
          ],
        )
      }).toThrow()
    })
  })

  it('should not allow arrays in single mode', () => {
    expect(() => {
      animation('rotate', ['move', '2s'], '1s')
    }).toThrow()
  })

  it('should not allow simple root level values in multi mode', () => {
    expect(() => {
      animation(['move', '2s'], 'rotate', '2s')
    }).toThrow()
  })
})
