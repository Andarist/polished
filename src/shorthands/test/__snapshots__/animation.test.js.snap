// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animation multi mode should be fine with less than eight arguments 1`] = `
Object {
  "animation": "rotate 1s ease-in-out",
}
`;

exports[`animation multi mode should be fine with multiple animations 1`] = `
Object {
  "animation": "rotate 1s ease-in-out, colorchange 2s",
}
`;

exports[`animation multi mode should pass first eight arguments to the CSS in multi mode 1`] = `
Object {
  "animation": "rotate 1s ease-in-out 0.5s 5 reverse forwards paused",
}
`;

exports[`animation single mode should be fine with less than eight arguments 1`] = `
Object {
  "animation": "rotate, 1s, ease-in-out",
}
`;

exports[`animation single mode should pass first eight arguments to the CSS 1`] = `
Object {
  "animation": "rotate, 1s, ease-in-out, 0.5s, 5, reverse, forwards, paused",
}
`;

exports[`animation single mode should throw a warning if more than eight elements are supplied 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- shorthands/animation.js --%c

animation %cexpects a maximum of %c1 parameter%c. However, you passed %c9 parameters%c. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#animation%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`animation single mode should throw an error if more than eight elements are supplied 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- shorthands/animation.js --%c

animation %cexpects a maximum of %c1 parameter%c. However, you passed %c9 parameters%c. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#animation%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
